### PR TITLE
Optimize concurrent gauges to use a shared current-time-in-minutes for all instances which use the system clock

### LINKED ISSUE #3398: Change ConcurrentGauge to use new PeriodicExecutor
## Environment Details
* Helidon Version: 2.x
* Helidon SE or Helidon MP SE
* JDK version:
* OS:
* Docker version (if applicable):

----------

## Problem Description
`ConcurrentGauge` depends on statistics reflecting the previous clock minute. The implementation currently computes the current minute from the current system time in ms on every update to the metric. This  can be costly under heavy load.

As part of PR #3385 which addresses issue #3389, a new class is added to `metrics/metrics` which `ConcurrentGauge` should use to optimize the calculation of current minute.

### PR DESCRIPTION
**Pull request by tjquinno**
Resolves #3398 

Earlier changes added a `PeriodicExecutor` which is used in metrics to optimize updates to current-time-in-seconds values.

This PR adds the use of `PeriodicExecutor` by the `ConcurrentGauge`  implementation which relies on a current-time-in-minutes value.

In real apps, virtually all `ConcurrentGauges` will use the system clock which means they can all share the same current-time-in-minutes value, which further means we can enroll only a single updater with `PeriodicExecutor`. 

This PR adds that optimization to `ConcurrentGauge` while preserving the existing behavior (which updates the current-time-in-minutes on each metric update) if a given instance _does not_ use the system clock.

Signed-off-by: tim.quinn@oracle.com <tim.quinn@oracle.com>

### CODE REVIEW

File: metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java [hunk 1]
```diff
@@ -183,17 +210,31 @@ public synchronized void dec() {
         }
 
         public synchronized void updateState() {
-            long currentMinute = currentTimeMinute();
-            long diff = currentMinute - lastMinute.get();
+            if (!usingSystemClock) {
+                updateCurrentMinute();
+            }
+            long diff = currentMinute.get() - lastMinute.get();
```
Comment (ljnelson) [inline metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java:216]: Not sure this has to be in the `synchronized` block, right?

File: metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java [hunk 2]
```diff
@@ -183,17 +210,31 @@ public synchronized void dec() {
...
             }
         }
 
         private long currentTimeMinute() {
-            return clock.milliTime() / 1000 / 60;
+            return currentTimeMinute(clock);
+        }
+
+        private static long currentTimeMinute(Clock clock) {
+            return clock.milliTime() / 1000 / 10;
```
Comment (olotenko) [inline metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java:229]: Is this deliberately in 10-seconds, not minutes?

File: metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java [hunk 3]
```diff
@@ -183,17 +210,31 @@ public synchronized void dec() {
...
+        }
+
+        private long updateCurrentMinute() {
+            return currentMinute.getAndSet(currentTimeMinute());
```
Comment (olotenko) [inline metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java:233]: It seems no one consumes the value returned by this `getAndSet`. Use `set`. In this context `lazySet` is also fine.

File: metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java [hunk 4]
```diff
@@ -183,17 +210,31 @@ public synchronized void dec() {
         }
 
         public synchronized void updateState() {
```
Comment (olotenko) [inline metrics/metrics/src/main/java/io/helidon/metrics/HelidonConcurrentGauge.java:212]: I do not quite get the point of `synchronized` here and in both `inc` and `dec`.

`synchronized` `inc` makes sense only if all accesses to `count` and `currentMax` are also `synchronized`, including reads, in which case we don't need them to be `AtomicLong`. But I presume they are `AtomicLong` to enable atomic updates that are cheaper than `synchronized`. In which case we don't need `synchronized`.

Atomic updates should be like so:

```
public void inc() {
   long count = count.incrementAndGet();
   if (count > currentMax.get()) currentMax.getAndUpdate(v -> count > v ? count: v);
}
```

`updateState` should probably be done only by the minute-updating routine.

And to check the intention: the `currentMin` and `currentMax` don't get reset after every minute intentionally?

**Review by olotenko — CHANGES_REQUESTED** [review]

**Comment by tjquinno** [comment]
Thanks to the reviewers, but...

I am closing this PR without merging, at least for now.

A test in the TCK schedules some updates to a `ConcurrentGauge` at precisely the top of a minute. The scheme in this PR for updating the shared current-time-in-minutes does not run (most likely) until after those updates to the metric, so the update logic incorrectly sees the update as happening in the previous minute, not the current one.

IMO, ideally the TCK test would schedule its work for 1 ms _after_ the start of the new minute. That way, I could revise the update scheme to schedule its update for exactly the start of each minute. For use cases other than the edge case where an update is scheduled at precisely the top of a minute, that would seem sufficient. 

But that would require a change in the TCK test, and MP might be heading towards adopting Micrometer for its metrics solution going forward so it seems fruitless to file an issue against the TCK under the circumstances.

Therefore, for now at least we'll just have to tolerate the performance hit from computing the current-time-in-minutes on each update.
